### PR TITLE
Performance optimizations for JPEG decompression

### PR DESCRIPTION
**Pull request by xqms**
This PR reduces JPEG decompression time by roughly factor 2 for larger images.

First of all, we use the TurboJPEG library directly rather than through OpenCV's imdecode() method. TurboJPEG is part of libjpeg-turbo and thus installed on most systems anyway. Using TurboJPEG has two advantages:

 * RGB/BGR conversions can be fused with the YUV to RGB conversion that is required during JPEG decompression anyway.
 * TurboJPEG's API makes it easy to achieve zero-copy operation.

We remove two data copies: One during RGB/BGR conversion and one during the final conversion to `sensor_msgs::Image`.

On my system, this PR allows me to decode a 4K JPEG stream (encoded externally from ROS) at 30 Hz without a problem. Without the PR, I could at most achieve 20 Hz.

I tested mostly the stream configuration I had available. I'll do further RGB/BGR tests soon. Since there are no unit tests for `compressed_image_transport` (or I didn't find them), this requires careful attention.

### CODE REVIEW

**Comment by peci1** [comment]
This PR breaks ABI by adding the `tjhandle tj_` member variable. Would there be an ABI-compatible way to implement this?

Otherwise, I agree a faster decode engine would be useful - currently, displaying a single 1600x7000 image stream in rviz on Steam Deck results in rviz running with 2 FPS, which is practically useless.

**Comment by xqms** [comment]
Hmm, good point. But is ABI a practical concern here? I would think most people use `compressed_image_transport` as a plugin and in that case the plugin base class is the externally visible ABI.

Keeping the ABI of the `CompressedImageSubscriber` class the same is very difficult, since the destructor is defined in the header - we have no possibility of doing any cleanup (e.g. freeing the `tj_` handle) on destruction. With a lot of hacking we could maybe change the generated code for the reconfigure class to stash away the handle there... But that seems very fragile.

So I would propose a solution in the middle: We create a new class (FastCompressedImageSubscriber or similar), which uses turbojpeg. The plugin XML files point to this new class. Anyone who is using the library `compressed_image_subscriber` directly (again, who would do that?) would still use the old class.

I don't have a lot of time to work on this, though. For my use cases I have a much-more performant alternative now (https://github.com/xqms/imgui_ros - documentation is coming!), so the pressing need for this on my side is now gone ;) I'm happy to make one more revision of this PR, though - or help anyone else who would be willing to take this up.

**Comment by peci1** [comment]
> But is ABI a practical concern here? I would think most people use `compressed_image_transport` as a plugin and in that case the plugin base class is the externally visible ABI.

Good point. I've actually tried to use the compressed transport as a direct library, but I haven't succeded - the API is just useless for any other use-case than `ImageSubscriber`. The decode/encode functions themselves are not factored out from the publishing code, so the idea to just use the decoder's C++ API did not work.

So maybe it is actually ok here to break ABI. Sorry for the confusion.

**Comment by ijnek** [comment]
I recently took over maintenance for the ROS 2 branches, but the maintainer for ROS 1 branches have been out for many years. I'm happy to merge this in if @peci1 or someone else could review.

File: compressed_image_transport/CMakeLists.txt
Comment (peci1) [inline compressed_image_transport/CMakeLists.txt]:
```suggestion
find_package(OpenCV REQUIRED)
find_package(catkin REQUIRED COMPONENTS cv_bridge dynamic_reconfigure image_transport)

find_package(PkgConfig REQUIRED)
pkg_check_modules(TurboJPEG REQUIRED libturbojpeg)
```

Comment (peci1) [reply compressed_image_transport/CMakeLists.txt]: Ubuntu, Fedora, Debian, Arch and probably all other distros provide a pkgconfig file, so let's use it.

File: compressed_image_transport/modules/FindTurboJPEG.cmake
Comment (peci1) [inline compressed_image_transport/modules/FindTurboJPEG.cmake]:
```suggestion
```
Not needed (substituted by pkg_config).

File: compressed_image_transport/include/compressed_image_transport/compressed_subscriber.h
Comment (peci1) [inline compressed_image_transport/include/compressed_image_transport/compressed_subscriber.h:74]: I tested performance without the class member (initing decompressor in each callback) and the performance loss wasn't that big. So if we'd want to keep ABI, this could be also a possibility.

File: compressed_image_transport/package.xml
Comment (peci1) [inline compressed_image_transport/package.xml]: I suggest switchin to package format 2 and declaring this only as `<exec_depend>` - otherwise all dependent packages would need to link to turbojpeg, which is not needed.

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
  // Parse format field
  std::string image_encoding;
  std::string compressed_encoding;
  {
    const size_t split_pos = message->format.find(';');
    if (split_pos != std::string::npos)
    {
      image_encoding = message->format.substr(0, split_pos);
      compressed_encoding = message->format.substr(split_pos);
    }
  }

  // Try TurboJPEG first (if the first bytes look like JPEG)
  if(message->data.size() > 4 && message->data[0] == 0xFF && message->data[1] == 0xD8)
  {
    sensor_msgs::ImagePtr decoded = decompressJPEG(message->data, image_encoding, message->header);
    if(decoded)
    {
      user_cb(decoded);
      return;
    }
  }

  // Otherwise, try our luck with OpenCV.
```

Do not unnecessarily rename existing variables.

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
    if (image_encoding.empty())
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
      cv_ptr->encoding = image_encoding;
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
      if ( enc::isColor(image_encoding))
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
        bool compressed_bgr_image = (compressed_encoding.find("compressed bgr") != std::string::npos);
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
          if ((image_encoding == enc::RGB8) || (image_encoding == enc::RGB16))
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
          if ((image_encoding == enc::RGBA8) || (image_encoding == enc::RGBA16))
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
          if ((image_encoding == enc::BGRA8) || (image_encoding == enc::BGRA16))
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
          if ((image_encoding == enc::BGR8) || (image_encoding == enc::BGR16))
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
          if ((image_encoding == enc::BGRA8) || (image_encoding == enc::BGRA16))
```

File: compressed_image_transport/src/compressed_subscriber.cpp
Comment (peci1) [inline compressed_image_transport/src/compressed_subscriber.cpp]:
```suggestion
          if ((image_encoding == enc::RGBA8) || (image_encoding == enc::RGBA16))
```

**Review by peci1 — CHANGES_REQUESTED** [review]
I went through this PR and locally tested it.

It works well on BGR, MONO8 and DEPTH (MONO16 PNG) images - the decoded output is the same as in the original version, and performance is better (except for the PNG images, where it is the same).

In my case, I tested performance with images 1616x7392 @ 6 Hz (a tall 6-camera merge from a synchronized omnicam). Playing with RTF set to 1.0, CPU usage of republish node decreases from ~60% to ~50%. When playing with RTF 10.0, the republish node manages republishing ~15 images/s, while the original implementation manages ~8 images/s.

I have some code comments that I'd like to be addressed. @xqms are you willing to still work on this PR, or should I take over it?

File: compressed_image_transport/CMakeLists.txt
Comment (xqms) [reply compressed_image_transport/CMakeLists.txt]: Ubuntu 16.04 does not provide a pkgconfig file. But I guess since Kinetic is EOL we don't care about 16.04.

File: compressed_image_transport/include/compressed_image_transport/compressed_subscriber.h
Comment (xqms) [reply compressed_image_transport/include/compressed_image_transport/compressed_subscriber.h:74]: I guess that is up to you. I still think nobody is using `compressed_image_transport` directly, so there should be no impact of an ABI change. But I'm also fine with moving it to the callback.

File: compressed_image_transport/package.xml
Comment (xqms) [reply compressed_image_transport/package.xml]: Good point, will do so.

Comment (xqms) [reply compressed_image_transport/package.xml]: Actually, at the moment we `#include <turbojpeg.h>` in `compressed_subscriber.h`, which makes it a build export dependency, right?

We could maybe replace the `tjhandle` with a `void*` and avoid the include in the header. Or have a pimpl-style `std::unique_ptr<Private>` member, that way we can define the `tjhandle` in the cpp file.

Comment (peci1) [reply compressed_image_transport/package.xml]: I think the reasoning here is the same as with the ABI compatibility - technically, it would add a build-export dependency if someone linked to the plugin, but doing so is highly improbable...

File: compressed_image_transport/include/compressed_image_transport/compressed_subscriber.h
Comment (peci1) [reply compressed_image_transport/include/compressed_image_transport/compressed_subscriber.h:74]: Feel free to leave it as it is. I think ABI compatibility here is not a key aspect as we're talking about a plugin.

File: compressed_image_transport/package.xml
Comment (xqms) [reply compressed_image_transport/package.xml]: Ok!

**Comment by peci1** [comment]
In the meantime, I've discovered a transport plugin doing the encoding part via turbojpeg: https://github.com/wep21/turbojpeg_compressed_image_transport/blob/main/src/turbojpeg_compressed_publisher.cpp . Would you be interested in doing that part, too? Maybe in a followup PR?

**Comment by xqms** [comment]
> In the meantime, I've discovered a transport plugin doing the encoding part via turbojpeg: https://github.com/wep21/turbojpeg_compressed_image_transport/blob/main/src/turbojpeg_compressed_publisher.cpp . Would you be interested in doing that part, too? Maybe in a followup PR?

Hmm, I currently don't have high-bandwidth JPEG publisher using `image_transport` in use, so this would be quite low on my priority list. I may get around to doing it sometime next year.

**Review by peci1 — APPROVED** [review]
I've retested this PR locally and everything works.

All my comments have been addressed, except for a dangling empty file `modules/FindTurboJPEG.cmake`, which is a remnant of the previous version and should be completely deleted.

**Comment by ijnek** [comment]
With the approval, I'm going to go ahead and merge this. Thanks for the contribution @xqms, sorry for the very long wait and thanks for reviewing @peci1!

**Comment by peci1** [comment]
Awesome! Thanks @xqms for working on this!

Could I also ask for a Melodic backport?

**Comment by ijnek** [comment]
Just realized the melodic version of image_transport_plugins points to the indigo-devel branch :open_mouth: 

I'm going to guess that the addition of libturbojpeg may cause issues in some users if they don't have libturbojpeg available on their system. So I'm hesitant for this backport. What do you think @peci1 ?

**Comment by peci1** [comment]
> points to the indigo-devel branch

That's quite common. There weren't too many breaking changes between indigo and melodic, so many packages share the same codebase for both. However, indigo is EOL for quite long, so I don't see a problem there. Regarding melodic, the situation is the same as for noetic - if somebody does not have libjpeg-turbo installed, he will need to add this dependency. However, that is also a common case. APT will take care of that in case the user runs on the binary packages. In case of from-source builds, the users should anyways do `rosdep install` after each source update, so this case should also be covered. And, last, libjpeg-turbo is available for all platforms of interest. I've personally checked this PR on melodic and it works fine.

**Comment by ijnek** [comment]
I'm comfortable with how things are done in ROS 2, but not too much ROS 1. Assuming things are done the same way, you'd never want to backport changes that add extra dependencies, even if you can confirm it works on the listed supported platforms. That way it caters for those on platforms that aren't listed but have successfully used the package up until the change.

That said, as you said libturbojpeg does seem covered pretty well across different distributions:
https://github.com/ros/rosdistro/blob/9cfaef7339b94ee28daa8ed1af69be870a37ab0d/rosdep/base.yaml#L5837-L5861

---

In regards to the indigo-devel branch, would you agree with branching off melodic-devel from indigo-devel and applying the changes there so we can ensure indigo-devel stays in a working state.

Looking forward to hearing your thoughts.

**Comment by peci1** [comment]
> you'd never want to backport changes that add extra dependencies

But this happens regularly when I update my system. Moreover, I don't understand the difference between noetic and melodic in this regard. Why would introducing a new dependency be a problem in Melodic and not a problem in Noetic?

> branching off melodic-devel

This might make sense as some distributions for kinetic do not have the PkgConfig for libjpeg-turbo. However, I'm not sure about the policy when a devel branch should be branched off, so that's probably more a maintainer-level decision.

**Comment by peci1** [comment]
A recent example of a dependency being added to Moveit "in the middle" of a released ROS version:

https://github.com/ros-planning/moveit/pull/2547/files#diff-a572b9bd78f220384c2c0ede5c92744f22f3bcc7bb6962847725f6507ef22cd4

and being backported to melodic here: https://github.com/ros-planning/moveit/pull/2651 .

**Comment by ijnek** [comment]
> Why would introducing a new dependency be a problem in Melodic and not a problem in Noetic?

This argument applies to merging in the changes to the noetic branch too. I just only thought of it after being requested to backport.

Given the links to the examples in moveit, I'm leaning towards branching-off melodic-devel and backporting the changes there as requested. Does that sound fine?

**Comment by peci1** [comment]
> branching-off melodic-devel and backporting the changes there as requested. Does that sound fine?

Perfect! Thanks.

**Comment by ijnek** [comment]
@Mergifyio backport melodic-devel

**Comment by peci1** [comment]
Looks promising.

---

Ahh, now I see there is still the empty cmake file in this PR, could you please remove it, too?

**Comment by ijnek** [comment]
Updating rosdistro for melodic here: https://github.com/ros/rosdistro/pull/35733

**Comment by ijnek** [comment]
> Ahh, now I see there is still the empty cmake file in this PR, could you please remove it, too?

https://github.com/ros-perception/image_transport_plugins/pull/124

**Pull request merged**
